build_runner_core/lib/src/package_graph/build_triggers.dart

```diff
@@ -0,0 +1,201 @@
+  static (Iterable<BuildTrigger>, List<String>) parseList({
+    required Object triggers,
+  }) {
+    if (triggers is! List<Object?>) {
+      return (
+        [],
+        ['Invalid `triggers`, should be a list of triggers: $triggers'],
+      );
+    }
+    final result = <BuildTrigger>[];
+    final warnings = <String>[];
+    for (final triggerString in triggers) {
+      BuildTrigger? trigger;
+      String? warning;
+      if (triggerString is String) {
+        (trigger, warning) = BuildTrigger.tryParse(triggerString);
+      }
+      if (trigger != null) {
+        result.add(trigger);
+      }
+      if (warning != null) {
+        warnings.add(warning);
+      }
+    }
+    return (result, warnings);
+  }
+
+  String get renderWarnings {
+    final result = StringBuffer();
+    for (final package in warningsByPackage.keys) {
+      result.writeln('build.yaml of package:$package:');
+      for (final warning in warningsByPackage[package]!) {
+        result.writeln('  $warning');
+      }
+    }
+    result.writeln(
+      'See https://pub.dev/packages/build_config#triggers for valid usage.',
+    );
+    return result.toString();
+  }
+}
+
+/// A build trigger: a heuristic that possibly skips running a build step based
+/// on the primary input source.
+abstract class BuildTrigger {
+  /// Parses a trigger, returning `null` on failure, optionally with a warning
+  /// message.
+  ///
+  /// The only supported trigger is [ImportBuildTrigger].
+  static (BuildTrigger?, String?) tryParse(String trigger) {
+    if (trigger.startsWith('import ')) {
+      trigger = trigger.substring('import '.length);
+      final result = ImportBuildTrigger(trigger);
+      return (result, result.warning);
+    } else if (trigger.startsWith('annotation ')) {
+      trigger = trigger.substring('annotation '.length);
+      final result = AnnotationBuildTrigger(trigger);
+      return (result, result.warning);
+    }
+    return (null, 'Invalid trigger: `$trigger`');
+  }
+
+  bool triggersOnPrimaryInput(String source);
+}
+
+// Note: `built_value` generated toString is relied on for digests, and
+// equality for testing.
+
+/// A build trigger that checks for a direct import.
+///
+/// TODO(davidmorgan): this currently over-matches, it just checks for a
+/// matching substring. It's a heuristic so over matching is okay, but it might
+/// be better to actually parse directives and do a correct match because
+/// directives are needed when resolving source.
+abstract class ImportBuildTrigger
+    implements
+        Built<ImportBuildTrigger, ImportBuildTriggerBuilder>,
+        BuildTrigger {
+  static final RegExp _regexp = RegExp(r'^[a-z][a-z0-9_/.]*$');
+  String get import;
+
+  ImportBuildTrigger._();
+  factory ImportBuildTrigger(String import) =>
+      _$ImportBuildTrigger._(import: import);
+
+  @override
+  bool triggersOnPrimaryInput(String source) {
+    return source.contains('package:$import');
```

Comment:
Technically an import string can be split into multiple lines, e.g.

```dart
import 'd' /* hello */ 'a'
//there
'r'
// I'm
't'
// a
':'
// perfectly
'i'
// valid
'o'
// import
''
// !
;
```

(and I have seen than --- not as absurdly as I did above but as I recall to make it fit on lines or something)